adafruit_mcp9600.py

```diff
@@ -0,0 +1,153 @@
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_MCP9600.git"
+from micropython import const
+from adafruit_bus_device.i2c_device import I2CDevice
+
+_DEFAULT_ADDRESS = const(0x67)
+
+_REGISTER_HOT_JUNCTION = const(0x00)
+_REGISTER_DELTA_TEMP = const(0x01)
+_REGISTER_COLD_JUNCTION = const(0x02)
+_REGISTER_THERM_CFG = const(0x05)
+_REGISTER_VERSION = const(0x20)
+
+class MCP9600():
+    """Interface to the MCP9600 thermocouple amplifier breakout"""
+
+    def __init__(self, i2c, address=_DEFAULT_ADDRESS, tctype="K", tcfilter=0):
+        types = ["K", "J", "T", "N", "S", "E", "B", "R"]
+        self.buf = bytearray(3)
+        self.i2c_device = I2CDevice(i2c, address)
+        self.type = tctype
+        # is this a valid thermocouple type?
+        if tctype not in types:
+            raise Exception("invalid thermocouple type ({})".format(tctype))
+        # filter is from 0 (none) to 7 (max), can limit spikes in
+        # temperature readings
+        tcfilter = min(7, max(0, tcfilter))
+        ttype = 0  # default is Type K
+        if tctype == "J":
+            ttype = 0x01
+        elif tctype == "T":
+            ttype = 0x02
+        elif tctype == "N":
+            ttype = 0x03
+        elif tctype == "S":
+            ttype = 0x04
+        elif tctype == "E":
+            ttype = 0x05
+        elif tctype == "B":
+            ttype = 0x06
+        elif tctype == "R":
+            ttype = 0x07
+
+        self.buf[0] = _REGISTER_THERM_CFG
+        self.buf[1] = tcfilter | (ttype << 4)
+        with self.i2c_device as tci2c:
+            tci2c.write(self.buf, end=2)
+
+    @property
+    def version(self):
+        """ MCP9600 chip version """
+        version = self._read_register(_REGISTER_VERSION, 1)
+        return version
+
+    @property
+    def ambient_temperature(self):
+        """ Cold junction/ambient/room temperature in Celsius """
+        data = self._read_register(_REGISTER_COLD_JUNCTION, 2)
+        if data[0] & 0xf0:
+            # negative temperature
+            value = data[0]*16.0 + data[1]/16.0 - 4096
+        else:
+            # positive temperature
+            value = data[0]*16.0 + data[1]/16.0
+        return value
+
+    @property
+    def hot_junction_temperature(self):
+        """ Hot junction temperature in Celsius """
+        data = self._read_register(_REGISTER_HOT_JUNCTION, 2)
+        if data[0] & 0x80:
+            # negative temperature
+            value = data[0]*16.0 + data[1]/16.0 - 4096
+        else:
+            # positive temperature
+            value = data[0]*16.0 + data[1]/16.0
+        return value
+
+    @property
+    def delta_temperature(self):
+        """ Delta temperature in Celsius """
+        data = self._read_register(_REGISTER_DELTA_TEMP, 2)
+        if data[0] & 0x80:
+            # negative temperature
+            value = data[0]*16.0 + data[1]/16.0 - 4096
+        else:
+            # positive temperature
+            value = data[0]*16.0 + data[1]/16.0
```

Comment:
Same here for using `struct.unpack`

Reply:
I simplified the return of _read_register() to return the entire array. Do you still think unpack is needed?

Reply:
I would suggest switching to using unpack. No need to re-invent the wheel, and it's a bit more succinct. Personally I trust it more than I trust my own ability to get the math right every time.

Your call, but I think it has its merits.

Reply:
So the code below uses the same algorithm as the [Arduino MCP9600 driver](https://github.com/adafruit/Adafruit_MCP9600/blob/master/Adafruit_MCP9600.cpp), using CP and unpack. Is this any better? The results are the same.

[(https://github.com/adafruit/Adafruit_MCP9600/blob/master/Adafruit_MCP9600.cpp)]


```
        data = self._read_register(_REGISTER_DELTA_TEMP, 2)
        d = unpack(">xH",data)
        if d[0] & 0x80:
            # negative temperature
            value = d[0]*0.0625 - 4096
        else:
            # positive temperature
            value = d[0]*0.0625
        return value
```

Reply:
Also, does CP care about big-endian and little-endian integers or does it abstract it away? I had to use big-endian for the unpack on the PyPortal, don't know if other CP boards will be different. The original algorithm does not care.

Reply:
So thinking about this the return data will always be big-endian since the data is always big-endian when returned from read_register(). So the original question still stands, which algorithm is better?

Reply:
Using `struct.unpack` is better. It's not available in Arduino, so that driver does it manually.

Your `_read_register` puts the actual read results in `buf[1]` and `buf[2]`. So you should be able to do something like:
```python
data = self._read_register(_REGISTER_HOT_JUNCTION, 2)
temperature = 0.0625 * struct.unpack(">h", data[1:])
```
Note how you can indicate endian-ness to `struct.unpack` with `<` or `>`.